specs/advanced-data-layouts/prollytree/spec.md

```diff
@@ -0,0 +1,467 @@
+
+### `ChunkConfig.maxChunkSize`
+
+The maximum size a chunk could be before it needs to be split regardless of the chunk boundries
+
+### `ChunkConfig.chunkStrategy`
+
+The string representing the type of strategy to use. Either `RollingHash` or  `KeySplitter`.
+	
+### `PrefixThresholdConfig`
+
+Config for the `PrefixThreshold` chunking strategy.
+
+This is the original strategy that was described in the Merkle Search Tree paper and has a `chunkingFactor` which represents the number of bits in the key which need to be `0` in order to set another boundry.
+
+Lower values result in wider nodes.
+
+### `WeibullThresholdConfig`
+
+Config for the `WeiBullThreshold` chunking strategy.
+
+Makes chunk boundary decisions on the hash of the key of a []byte pair and tries to create chunks that have an average number of []byte pairs, rather than an average number of Bytes. However, because the target number of []byte pairs is computed directly from the chunk size and count, the practical difference in the distribution of chunk sizes is minimal. It uses a dynamic threshold modeled on a weibull distribution (https://en.wikipedia.org/wiki/Weibull_distribution). As the size of the current trunk increases, it becomes easier to pass the threshold, reducing the likelihood of forming very large or very small chunks.
+
+The `K` value represents the Shape Parameter, and `L` represents `λ` - the scaling factor.
+
+### `RollingHashConfig`
+
+ rollingHashSplitter is a nodeSplitter that makes chunk boundary decisions using a rolling value hasher that processes Item pairs in a byte-wise fashion.
+
+rollingHashSplitter uses a dynamic hash pattern designed to constrain the chunk Size distribution by reducing the likelihood of forming very large or very small chunks. As the Size of the current chunk grows, rollingHashSplitter changes the target pattern to make it easier to match. The result is a chunk Size distribution that is closer to a binomial distribution, rather than geometric.
+
+## Algorithm in detail
+
+Left to the implementation is how to generate and load data from IPLD.
+Whenever a Link is resolved to a ProllyNode or ChunkConfig, this is done via the IPLD LinkSystem which it an implementation detail outside of the scope of this document.
+
+When serializing data into a CID+Block, one should use the codec and multihash that's either used by root CID of the tree you are modifying, or specificed explicitly during tree creation. Blocks should also be saved to the IPLD LinkSystem and how this is done is also outside the scope of this document. We will assume that there is a `getNode(cid)` API which loads IPLD Nodes from a CID, and a `saveNode(node) => {cid, bytes}` which will save a node and get back a CID and the byte contents used for generating the CID. Note that `saveNode` should use the same encoding and hash algorithm as the root of the tree.
+
+### IsLeaf(ProllyNode) : Boolean
```

Comment:
the functions(include below) may be defined as structure functions? such as func(n *ProllyNode) IsLeaf() bool

Reply:
I was thinking we could purposefully ignore any Object Oriented bits to the code so that it'd make as few assumptions as possible about the programming environment. Structure functions are very golang-y so it might be more specific than necessary for the spec.

Implementers can use these function signatures as vague guides they can use to write their code but are free to implement whatever function signatures they need. The important part is the shape of the data and making sure they consistently create the same trees from the same input.